### ISSUE TITLE
Type checker gives false positives - Mismatched types [E0308]

### ISSUE THREAD

**Issue by mouse07410** [intellij-rust/intellij-rust#3882]
## Environment

* **Intellij-Rust plugin version:** v0.2.97.2123-191
* **Rust toolchain version:** stable-x86_64-apple-darwin (default), rustc 1.34.2 (6c2484dc3 2019-05-13)
* **IDE name and version:** CLion 2019.1.3 Build #CL-191.7141.37, build on May 6, 2019
* **Operating system:** MacOS Mojave 10.14.4

## Problem description
In the code below, IntelliJ_Rust plugin erroneously complains about type mismatch in `CStr::from_ptr()` argument:
![Screen Shot 2019-05-22 at 2 36 28 PM](https://user-images.githubusercontent.com/5923577/58219593-843f5000-7cd9-11e9-8d1a-ba807e9fc7c6.png)

Neither `cargo check --all` nor `cargo clippy --all` find any problem with the code whatsoever:
```
$ cargo clippy --all
   Compiling ffi v0.1.0 (/Users/ur20980/src/rust-class-ffi)
    Finished dev [unoptimized + debuginfo] target(s) in 0.86s
$ cargo check --all
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
$ 
```

## Steps to reproduce

Load the project, observe the errors the IDE displays.

Possibly the same problem as reported in #3514. I discovered working with FFI. 

Relevant Rust code:
```rust
use std::ffi::CStr;

mod ffi {
    use libc::{c_char, c_int};

    #[repr(C)]
    pub struct AwesomeType {
        _private: [i32; 0],
    }

    extern "C" {
        pub fn awesome_features_enabled() -> *const c_char;
    }
}

pub fn features_enabled() -> &'static str {
    unsafe {
        let ptr = ffi::awesome_features_enabled();
        assert!(!ptr.is_null());
        let cstr: &CStr = CStr::from_ptr(ptr);
        cstr.to_str().unwrap()
    }
}
```

Relevant C code:
```C
#include <stdlib.h>
#include <memory.h>
#include <string.h>
#include "awesome_library.h"

static char *FEATURES = "portability, speed";

char *awesome_features_enabled(void)
{
  return FEATURES;
}
```

**Comment by vlad20012** [comment]
There are multiple `c_char` definitions under `cfg` attributes.|
```rust
#[cfg(...)]
pub type c_char = u8;
#[cfg(...)]
pub type c_char = i8;
```
Unfortunately, we don't support `cfg` attributes for now

**Comment by mouse07410** [comment]
Understood.

The code compiles and runs regardless, which is good.

Is there a workaround to silence this error, but keep Type checker enabled?

**Comment by vlad20012** [comment]
> Also, what file contains those attributes?
It's std's `c_char` definition: `/libstd/os/raw/mod.rs:29`

> Is there a workaround to silence this error, but keep Type checker enabled?
You can disable our typechecker and enable Cargo Check linter (if it's fast enough for your project) or suppress typechecker inspection for some function

**Comment by mouse07410** [comment]
Understood, thanks.

1. Could you provide instructions how to ensure Cargo Check linter?  
2. At what point is your linter activated - when/how often is the code scanned?
3. How to disable your typechecker?
4. How to suppress your typechecker for a specific function?

**Comment by vlad20012** [comment]
1. ctrl+shift+A "cargo check" or `Settings | Languages & Frameworks | Rust | Cargo | Rust external linter to analyze code on the fly`
2. See `Settings | Editor | General | Autoreparse delay` ([docs](https://www.jetbrains.com/help/idea/settings-editor-general.html)). Same with Cargo Check
3. 
![Peek 2019-05-28 17-19](https://user-images.githubusercontent.com/3221931/58485460-c4d11a80-816c-11e9-868f-b26cfe3adeb8.gif)
5. Of course! :)

**Comment by mouse07410** [comment]
Thank you! Your responsiveness is appreciated.

> `Autoreparse delay` 

Does it mean that if within the configured "delay time" there was no more text entered - the IDE will re-parse *only* the modified/entered text? 

> Same with Cargo Check

There's "Run external linter on the fly" option - but no info how often this "on the fly" is. How often/when would "cargo check" run, and how much code would it cover? Does enabling it means that every time I enter a new line, the whole file will be scanned via "cargo check"? Or...? 

If both Type Checker and External linter are enabled - will they run both? Or does the External linter take precedence?

Re. Disabling the Type Checker. It seems safer to localize the problem areas into reasonably small functions, and disable the Type Checker only for those functions. Until the appropriate capabilities are added.

> Of course! :)

Great! Might I suggest/request that this is done rather sooner than later?

Thanks!

**Comment by MarcSeebold** [comment]
Same here.
`
fn foo(bar: &u32) { bar + 1; }
`

**Comment by vlad20012** [comment]
@MarcSeebold can't reproduce it on the latest nightly plugin (if the only release affected, most likely that we already fixed it recently). Could you please check it on the latest nightly plugin (see [instructions](https://intellij-rust.github.io/install.html))

**Comment by mouse07410** [comment]
Just tested my project https://github.com/intellij-rust/intellij-rust/issues/3882#issue-447411490 against the nightly Rust (and TOML), got exactly the same result - it fails the Type Checker. So, no observed improvement in this between the stable and the nightly.

**Comment by vlad20012** [comment]
@mouse07410 my comment was addressed to @MarcSeebold, it's really another issue. And I talked about  nightly IntelliJ Rust plugin, not nightly rust

**Comment by mouse07410** [comment]
> they have another issue

Ah, understood. Sorry for the noise.

> And I talked about nightly IntelliJ Rust plugin

Understood. So was I.

**Comment by vlad20012** [comment]
Already fixed
